sparse/file.go

```diff
@@ -157,7 +157,7 @@ func alignmentShift(block []byte) int {
 }
 
 func ReadDataInterval(file FileIoProcessor, dataInterval Interval) ([]byte, error) {
-	data := make([]byte, dataInterval.Len())
+	data := AllocateAligned(int(dataInterval.Len()))
```

Comment:
Fair point, let me explain why I decide to do this here:
- allocating an aligned buffer is just as cheap as allocating len(unaligned buffer + 4K)
- this is since all this does is allocate a buffer + 4K then return a slice of that buffer where the base[0] points to an aligned memory address.

Now for the case where it really matters, directIO
- in the case where the randomly allocated buffer is unaligned:
-- we need to do 2 buffer allocations (2x the memory)
-- a full buffer copy after the read

Both of these operations are unnecessary, considering that we are most likely going to be using direct IO.
But even if not, we don't really incur much additional cost, but save costs for each ReadInterval call that is done with directIO (Data read, Checksum, etc)

------------------------------------------------------------
Summary:
In the case where we are using direction IO, we save an extra len(buf + 4K) allocation + memory copy.
In the case where we use buffered IO, we carry an extra (4K memory allocation) for the lifetime of that buffer.

What do you think? Any concerns?

Reply:
Thank you for the explanation. It's a really nice optimization!